drivers/net/ethernet/renesas/rswitch.c

```diff
@@ -3764,17 +3808,107 @@ static void rswitch_deinit_rdev(struct rswitch_private *priv, int index)
 
 static void rswitch_deinit(struct rswitch_private *priv)
 {
-	int i;
+	struct rswitch_device *rdev;
 
-	for (i = 0; i < num_ndev; i++) {
-		rswitch_deinit_rdev(priv, i);
-		rswitch_ndev_unregister(priv, i);
+	list_for_each_entry(rdev, &priv->rdev_list, list) {
+		rswitch_deinit_rdev(rdev);
+		rswitch_ndev_unregister(rdev);
 	}
 
 	rswitch_free_irqs(priv);
 	rswitch_desc_free(priv);
 }
 
+static int vlan_dev_register(struct net_device *dev)
+{
+	struct net_device *real_dev;
+	struct rswitch_net *rn;
+	struct rswitch_private *priv;
+	struct rswitch_device *rdev, *parent_rdev;
+	int ret;
+	rn = net_generic(&init_net, rswitch_net_id);
+	priv = rn->priv;
+
+	real_dev = vlan_dev_real_dev(dev);
+
+	if (!ndev_is_tsn_dev(real_dev, priv))
+		return 0;
+
+	parent_rdev = netdev_priv(real_dev);
+
+	rdev = kzalloc(sizeof(*rdev), GFP_KERNEL);
+	if (!rdev)
+		return -ENOMEM;
+	dev->dev.parent = real_dev->dev.parent;
+	rdev->ndev = dev;
+	rdev->priv = priv;
+	INIT_LIST_HEAD(&rdev->routing_list);
+	INIT_LIST_HEAD(&rdev->tc_u32_list);
+	INIT_LIST_HEAD(&rdev->tc_matchall_list);
+	INIT_LIST_HEAD(&rdev->tc_flower_list);
+	INIT_LIST_HEAD(&rdev->list);
+	rdev->port = -1;
+	rdev->etha = NULL;
+	rdev->addr = priv->addr;
+	spin_lock_init(&rdev->lock);
+	list_add(&rdev->list, &priv->rdev_list);
+
+	ret = rswitch_txdmac_init(dev, priv, -1);
+	if (ret)
+		goto err_tx;
+	ret = rswitch_rxdmac_init(dev, priv, -1);
+	if (ret)
+		goto err_rx;
+
+	netdev_info(dev, "MAC address %pMn", dev->dev_addr);
+	return 0;
+err_rx:
+	rswitch_txdmac_free(dev, priv);
+err_tx:
+	list_del(&rdev->list);
+	return ret;
+}
+
+static void vlan_dev_unregister(struct net_device *dev)
+{
+	struct rswitch_device *rdev;
+	struct rswitch_net *rn;
+	struct rswitch_private *priv;
+
+	rn = net_generic(&init_net, rswitch_net_id);
+	priv = rn->priv;
+	rdev = ndev_to_rdev(dev);
+	rswitch_rxdmac_free(dev, priv);
+	rswitch_txdmac_free(dev, priv);
+
+	list_del(&rdev->list);
+}
+
+static int vlan_device_event(struct notifier_block *unused, unsigned long event,
+			     void *ptr)
+{
+	struct net_device *dev = netdev_notifier_info_to_dev(ptr);
+
+	if (!is_vlan_dev(dev)) {
```

Comment:
Nit: Kernel coding [style](https://www.kernel.org/doc/html/v4.10/process/coding-style.html) says:
"Do not unnecessarily use braces where a single statement will do."
So, could you remove braces?